Close 'closed channel' cause panic

### Your environment.
  - Version: *pion/ice/v2@v2.2.3*
  - Browser: *Chrome newest version*
  - Other Information - *using simulcast feature*

### What did you do?
I'm trying to re-handshake peerconnection between `pion` and `chrome` continuously

### What happened?
I think the root cause is `gatherCandidateDone` channel was closed before but still to do it again.

<img width="648" alt="Screen Shot 2022-04-22 at 16 27 26" src="https://user-images.githubusercontent.com/31619361/164678158-6c1011ad-43e4-4e9c-9dca-905be7072787.png">

<img width="677" alt="Screen Shot 2022-04-22 at 16 21 44" src="https://user-images.githubusercontent.com/31619361/164677186-a6dc1a7d-8d6e-401b-8cad-cfe6ea0a2c12.png">

I am facing this very rare panic. Going to put the text here so that it's indexable for search:

```go
panic: close of closed channel
goroutine 47164 [running]:
github.com/pion/ice/v2.(*Agent).gatherCandidates(0xc00153cf00, {0x14740a8?, 0xc002e8d180})
	/go/pkg/mod/github.com/pion/ice/v2@v2.3.1/gather.go:112 +0x58e
created by github.com/pion/ice/v2.(*Agent).GatherCandidates.func1
	/go/pkg/mod/github.com/pion/ice/v2@v2.3.1/gather.go:56 +0x1db
```

My assumption is, that canceling previous gathering does not happen synchronously.
https://github.com/pion/ice/blob/be72bc709bb290aba9bdc842988d2f893300d967/gather.go#L51-L54

If it takes a little bit longer, it can close newly created `gatherCandidateDone` channel.
https://github.com/pion/ice/blob/be72bc709bb290aba9bdc842988d2f893300d967/gather.go#L64

We need to make sure previous gathering completed before starting new.

--- a/gather.go
+++ b/gather.go
@@ -51,17 +51,18 @@ func (a *Agent) GatherCandidates() error {
 		a.gatherCandidateCancel() // Cancel previous gathering routine
 		ctx, cancel := context.WithCancel(ctx)
 		a.gatherCandidateCancel = cancel
-		a.gatherCandidateDone = make(chan struct{})
+		done := make(chan struct{})
+		a.gatherCandidateDone = done
 
-		go a.gatherCandidates(ctx)
+		go a.gatherCandidates(ctx, done)
 	}); runErr != nil {
 		return runErr
 	}
 	return gatherErr
 }
 
-func (a *Agent) gatherCandidates(ctx context.Context) {
-	defer close(a.gatherCandidateDone)
+func (a *Agent) gatherCandidates(ctx context.Context, done chan struct{}) {
+	defer close(done)
 	if err := a.setGatheringState(GatheringStateGathering); err != nil { //nolint:contextcheck
 		a.log.Warnf("failed to set gatheringState to GatheringStateGathering: %v", err)
 		return

--- a/gather_test.go
+++ b/gather_test.go
@@ -87,6 +87,34 @@ func TestListenUDP(t *testing.T) {
 	assert.NoError(t, a.Close())
 }
 
+func TestGatherConcurrency(t *testing.T) {
+	report := test.CheckRoutines(t)
+	defer report()
+
+	lim := test.TimeOut(time.Second * 30)
+	defer lim.Stop()
+
+	a, err := NewAgent(&AgentConfig{
+		NetworkTypes:    []NetworkType{NetworkTypeUDP4, NetworkTypeUDP6},
+		IncludeLoopback: true,
+	})
+	assert.NoError(t, err)
+
+	candidateGathered, candidateGatheredFunc := context.WithCancel(context.Background())
+	assert.NoError(t, a.OnCandidate(func(c Candidate) {
+		candidateGatheredFunc()
+	}))
+
+	// tesing for panic
+	for i := 0; i < 10; i++ {
+		_ = a.GatherCandidates()
+	}
+
+	<-candidateGathered.Done()
+
+	assert.NoError(t, a.Close())
+}
+
 func TestLoopbackCandidate(t *testing.T) {
 	report := test.CheckRoutines(t)
 	defer report()

